File: Laba2/Laba2/ClassArray.cs
﻿using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Laba2
{
    class ClassArray<T> : /*IEnumerator<T>,*/ IEnumerable<T>, IComparable<ClassArray<T>>
    {
        private Dictionary<int,T> places;

        private int maxCount;

        private T defaultValue;

        public ClassArray(int size, T defVal)
        {
            defaultValue = defVal;
            places = new Dictionary<int, T>();
            maxCount = size;
        }

        public static int operator +(ClassArray<T> p, T loc)
        {
            var isHeatovoz = loc is Heatovoz;
            if (p.places.Count == p.maxCount)
            {
                throw new DepoOverflowException();
            }
            int index = p.places.Count;
            for (int i = 0; i < p.places.Count; ++i)
            {
                if (p.CheakFreePlace(i))
                {
                    index = i;
                }
                if (loc.GetType() == p.places[i].GetType())
                {
                    if (isHeatovoz)
                    {
                        if((loc as Heatovoz).Equals(p.places[i]))
                        {
                            throw new DepoAlreadyHaveException();
                        }
                        
                    }
                    else if ((loc as Locomotive).Equals(p.places[i]))
                    {
                        throw new DepoAlreadyHaveException();
                    }
                }
            }
            if (index != p.places.Count)
            {
                p.places.Add(index, loc);
                return index;
            }
            p.places.Add(p.places.Count, loc);
            return p.places.Count - 1;
        }

        public static T operator -(ClassArray<T> p, int index)
        {
            if (p.places.ContainsKey(index))
            {
                T car = p.places[index];
                p.places.Remove(index);
                return car;
            }

            throw new DepoIndexOutOfRangeException();
        }

        private bool CheakFreePlace(int index)
        {
            return !places.ContainsKey(index);

        }
        public T this[int ind]
        {
            get
            {
                if (places.ContainsKey(ind))
                {
                    return places[ind];
                }
                return defaultValue;
            }
        }

        /*private int currentIndex;

        public T Current
        {
            get
            {
                return places[places.Keys.ToList()[currentIndex]];
            }
        }

        object IEnumerator.Current
        {
            get
            {
                return Current;
            }
        }

        public void Dispose() { }

        public bool MoveNext()
        {
            if (currentIndex + 1 >= places.Count)
            {
                Reset();
                return false;
            }
            currentIndex++;
            return true;
        }
        public void Reset()
        {
            currentIndex = -1;
        }*/

        public IEnumerator<T> GetEnumerator()
        {
            return places.Values.GetEnumerator();
        }
        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }

        public int CompareTo(ClassArray<T> other)
        {
            if (this.Count() > other.Count())
            {
                return -1;
            }
            else if (this.Count() < other.Count())
            {
                return 1;
            }
            else
            {
                var thisKeys = places.Keys.ToList();
                var otherKeys = other.places.Keys.ToList();
                for(int i = 0; i < places.Count; ++i)
                {
                    if (places[thisKeys[i]]is Heatovoz && other.places[thisKeys[i]]is Locomotive)
                    {
                        return 1;
                    }
                    if (places[thisKeys[i]] is Locomotive && other.places[thisKeys[i]] is Heatovoz)
                    {
                        return -1;
                    }
                    if (places[thisKeys[i]] is Heatovoz && other.places[thisKeys[i]] is Heatovoz)
                    {
                        return (places[thisKeys[i]] is Heatovoz).CompareTo(other.places[thisKeys[i]] is Heatovoz);
                    }
                    if (places[thisKeys[i]] is Locomotive && other.places[thisKeys[i]] is Locomotive)
                    {
                        return (places[thisKeys[i]] is Locomotive).CompareTo(other.places[thisKeys[i]] is Locomotive);
                    }
                }
            }
            return 0;
        }
    }
}

File: Laba2/src/main/ClassArray.java
package main;

import java.io.*;
import java.util.*;

import components.*;
import train.Heatovoz;
import train.ITransport;
import train.Locomotive;

public class ClassArray<T extends ITransport> implements Externalizable, Comparable<ClassArray<T>>, Iterable<T>, Iterator<T> {
	
	private static final long serialVersionUID = 1L;
	
	private Map<Integer,T> places;
	
	private int maxCount;
	
	private T defaultValue;
	
	public ClassArray(int size, T defVal){
		defaultValue=defVal;
		places=new HashMap<>();
		maxCount=size;
	}
	public ClassArray(){}
	
	public T GetLoc(int index){
		if(places.containsKey(index)){
			return places.get(index);
		}
		return defaultValue;
	}
	
	public int Add(T loc) throws DepoOverflowException, DepoAlreadyHaveException {
		if(places.size()==maxCount) {
			throw new DepoOverflowException();
		}
		int index=places.size();
		
		for(int i=0;i<places.size();i++){
			if(CheakFreePlace(i)){
				index=i;
				
			}
			if(loc.getClass().equals(places.get(i).getClass())){
				if(loc instanceof Heatovoz){
					if(((Heatovoz)loc).equals(places.get(i))){
						throw new DepoAlreadyHaveException();
					}
				}else if(((Locomotive)loc).equals(places.get(i))){
					throw new DepoAlreadyHaveException();
				}
			}
		}
		if(index != places.size()){
			places.put(index, loc);
			return index;
		}
		places.put(places.size(), loc);
		return places.size()-1;
	}
	
	public T Get(int index) throws DepoIndexOutOfRangeException {
		if(places.containsKey(index)){
			return places.remove(index);
		}
		throw new DepoIndexOutOfRangeException();
	}
	
	private boolean CheakFreePlace(int index)
    {
        return !places.containsKey(index);

    }

	@Override
	public void writeExternal(ObjectOutput out) throws IOException {
		out.writeObject(places);
		
	}

	@Override
	public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
		places=(Map<Integer,T>)in.readObject();
		
	}

private int currentIndex = -1;
	
	public int size(){
		return places.size();
	}
	public Map<Integer, T> getPlaces() {
		return places;
	}

	@Override
	public int compareTo(ClassArray<T> other) {
		if (this.size() > other.size())
        {
            return -1;
        }
        else if (this.size() < other.size())
        {
            return 1;
        }
        else
        {
        	Integer[] thisKeys=places.keySet().toArray(new Integer[this.size()]);
        	Integer[] otherKeys=other.getPlaces().keySet().toArray(new Integer[other.size()]);
            for(int i=0;i<places.size();++i)
            {
                if(places.get(thisKeys[i]).getClass().equals(Heatovoz.class) && other.getPlaces().get(otherKeys[i]).getClass().equals(Locomotive.class)){
                	return 1;
                }
                if (places.get(thisKeys[i]).getClass().equals(Locomotive.class) && other.getPlaces().get(otherKeys[i]).getClass().equals(Heatovoz.class))
                {
                    return -1;
                }
                if (places.get(thisKeys[i]).getClass().equals(Heatovoz.class) && other.getPlaces().get(otherKeys[i]).getClass().equals(Heatovoz.class))
                {
                    return ((Heatovoz)places.get(thisKeys[i])).compareTo((Heatovoz)other.getPlaces().get(otherKeys[i]));
                }
                if (places.get(thisKeys[i]).getClass().equals(Locomotive.class) && other.getPlaces().get(otherKeys[i]).getClass().equals(Locomotive.class))
                {
                	return ((Locomotive)places.get(thisKeys[i])).compareTo((Locomotive)other.getPlaces().get(otherKeys[i]));
                }
            }
        }
		return 0;
	}

	private void Reset() {
		currentIndex = -1;
	}

	@Override
	public boolean hasNext() {
		if (currentIndex + 1 >= places.size()) {
			Reset();
			return false;
		}
		currentIndex++;
		return true;
	}

	@Override
	public T next() {
		return places.get(currentIndex);
	}

	@Override
	public Iterator<T> iterator() {
		return this;
	}
}
